packages/analytics-js/src/components/userSessionManager/UserSessionManager.ts

```diff
@@ -58,70 +78,119 @@ class UserSessionManager implements IUserSessionManager {
+    this.setGroupTraits(this.getGroupTraits() ?? defaultUserSessionValues.groupTraits);
+    this.setAnonymousId(this.getAnonymousId(state.loadOptions.value.anonymousIdOptions));
+
+    const initialReferrer = this.getInitialReferrer();
+    const initialReferringDomain = this.getInitialReferringDomain();
+
+    if (initialReferrer && initialReferringDomain) {
+      this.setInitialReferrer(initialReferrer);
+      this.setInitialReferringDomain(initialReferringDomain);
+    } else if (initialReferrer) {
+      this.setInitialReferrer(initialReferrer);
+      this.setInitialReferringDomain(getReferringDomain(initialReferrer));
+    } else {
+      const referrer = getReferrer();
+      this.setInitialReferrer(referrer);
+      this.setInitialReferringDomain(getReferringDomain(referrer));
+    }
+    // Initialize session tracking
+    if (this.isStorageTypeValidForStoringData(sessionTrackingStorage)) {
       this.initializeSessionTracking();
-      // Register the effect to sync with storage
-      this.registerEffects();
+    } else {
+      state.session.sessionInfo.value = defaultUserSessionValues.sessionInfo;
     }
+    // Register the effect to sync with storage
+    this.registerEffects();
   }
 
-  setDefaultValues() {
-    batch(() => {
-      state.session.userId.value = defaultUserSessionValues.userId;
-      state.session.userTraits.value = defaultUserSessionValues.userTraits;
-      state.session.groupId.value = defaultUserSessionValues.groupId;
-      state.session.groupTraits.value = defaultUserSessionValues.groupTraits;
-      state.session.anonymousUserId.value = defaultUserSessionValues.anonymousUserId;
-      state.session.initialReferrer.value = defaultUserSessionValues.initialReferrer;
-      state.session.initialReferringDomain.value = defaultUserSessionValues.initialReferringDomain;
-      state.session.sessionInfo.value = defaultUserSessionValues.sessionInfo;
+  isStorageTypeValidForStoringData(storage: StorageType): boolean {
+    return !!(
+      storage === COOKIE_STORAGE ||
+      storage === LOCAL_STORAGE ||
+      storage === MEMORY_STORAGE
+    );
+  }
+
+  migrateDataFromPreviousStorage() {
+    const entries: UserSessionEntries = state.storage.entries.value as UserSessionEntries;
+    Object.keys(entries).forEach(entry => {
+      const key = entry as UserSessionStorageKeysType;
+      const currentStorage = entries[key].storage;
+      const storages: MigrationStorageType[] = [COOKIE_STORAGE, LOCAL_STORAGE];
+
+      storages.forEach(storage => {
+        const store = this.storeManager?.getStore(storageClientDataStoreNameMap[storage]);
+        const curStore = this.storeManager?.getStore(storageClientDataStoreNameMap[currentStorage]);
+        if (
+          storage !== currentStorage &&
+          store &&
+          this.isStorageTypeValidForStoringData(currentStorage)
+        ) {
+          const value = store?.get(userSessionStorageKeys[key]);
+          if (value) {
```

Comment:
Did you mean the if checking?

Reply:
Yes.

Reply:
Say userId is in the `cookie` and current storage is also `cookie`.
In the loop when current storage is `cookie` and storage is `local storage` it will enter inside the loop. now it will try to fetch userId from `local storage` which is not there. Now if we remove the if checking it will set userId as null/nothing.
This is wrong.

Reply:
The correct thing to check for is empty strings, `null`, and `undefined` values.

Reply:
Done